cmd/snap/cmd_run.go

```diff
@@ -482,26 +491,39 @@ func (x *cmdRun) straceOpts() (opts []string, raw bool, err error) {
 	return opts, raw, nil
 }
 
+func getInfoAndApp(snapName, appName string, rev snap.Revision) (*snap.Info, *snap.AppInfo, error) {
+	info, err := getSnapInfo(snapName, rev)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	app := info.Apps[appName]
+	if app == nil {
+		return nil, nil, fmt.Errorf(i18n.G("cannot find app %q in %q"), appName, snapName)
+	}
+
+	return info, app, nil
+}
+
 func (x *cmdRun) snapRunApp(snapApp string, args []string) error {
 	if x.DebugLog {
 		os.Setenv("SNAPD_DEBUG", "1")
 		logger.Debugf("enabled debug logging of early snap startup")
 	}
 	snapName, appName := snap.SplitSnapApp(snapApp)
-	info, err := getSnapInfo(snapName, snap.R(0))
+
+	snapRev := snap.R(0)
+	hint, inhibitInfo, err := runinhibit.IsLocked(snapName)
 	if err != nil {
 		return err
 	}
-
-	app := info.Apps[appName]
-	if app == nil {
-		return fmt.Errorf(i18n.G("cannot find app %q in %q"), appName, snapName)
+	if hint != runinhibit.HintNotInhibited {
+		snapRev = inhibitInfo.Previous
```

Comment:
any reason not to wait already here in this approach? which would also mean maybe we don't need the changes to remember the snap revision? or do we have other places where we need that?

Reply:
no, maybe we want that logic, but the main point about needing to hold the lock to get the info with snap.R(0) and needing to wait/grab the lock earlier stands

Reply:
This is done later in runSnapConfine here: https://github.com/snapcore/snapd/pull/13411/files#diff-8362a8b80bc8ecb25f0a01a694035de268e569287891db10dfd0426c1a411f25R1062-R1076